hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequestV1.java

```diff
@@ -0,0 +1,275 @@
+    } catch (IOException ex) {
+      result = Result.FAILURE;
+      exception = ex;
+      omClientResponse = new OMKeyCommitResponseV1(createErrorOMResponse(
+              omResponse, exception));
+    } finally {
+      addResponseToDoubleBuffer(trxnLogIndex, omClientResponse,
+              omDoubleBufferHelper);
+
+      if(bucketLockAcquired) {
+        omMetadataManager.getLock().releaseWriteLock(BUCKET_LOCK, volumeName,
+                bucketName);
+      }
+    }
+
+    auditLog(auditLogger, buildAuditMessage(OMAction.COMMIT_KEY, auditMap,
+            exception, getOmRequest().getUserInfo()));
+
+    processResult(commitKeyRequest, volumeName, bucketName, keyName, omMetrics,
+            exception, omKeyInfo, result);
+
+    return omClientResponse;
+  }
+
+
+  /**
+   * Check for directory exists with same name, if it exists throw error.
+   *
+   * @param keyName                  key name
+   * @param ozoneManager             Ozone Manager
+   * @param reachedLastPathComponent true if the path component is a fileName
+   * @throws IOException if directory exists with same name
+   */
+  private void checkDirectoryAlreadyExists(String keyName,
+                                           OzoneManager ozoneManager,
+                                           boolean reachedLastPathComponent)
+          throws IOException {
+    // Reached last component, which would be a file. Returns its parentID.
+    if (reachedLastPathComponent && ozoneManager.getEnableFileSystemPaths()) {
+      throw new OMException("Can not create file: " + keyName +
+              " as there is already directory in the given path", NOT_A_FILE);
+    }
+  }
+
+  /**
+   * Get parent id for the user given path.
+   *
+   * @param bucketId          bucket id
+   * @param pathComponents    fie path elements
+   * @param keyName           user given key name
+   * @param omMetadataManager metadata manager
+   * @return lastKnownParentID
+   * @throws IOException DB failure or parent not exists in DirectoryTable
+   */
+  private long getParentID(long bucketId, Iterator<Path> pathComponents,
+                           String keyName, OMMetadataManager omMetadataManager,
+                           OzoneManager ozoneManager)
+          throws IOException {
+
+    long lastKnownParentId = bucketId;
+
+    // If no sub-dirs then bucketID is the root/parent.
+    if(!pathComponents.hasNext()){
+      return bucketId;
+    }
+
+    OmDirectoryInfo omDirectoryInfo;
+    while (pathComponents.hasNext()) {
+      String nodeName = pathComponents.next().toString();
+      boolean reachedLastPathComponent = !pathComponents.hasNext();
+      String dbNodeName =
+              omMetadataManager.getOzonePathKey(lastKnownParentId, nodeName);
+
+      omDirectoryInfo = omMetadataManager.
+              getDirectoryTable().get(dbNodeName);
+      if (omDirectoryInfo != null) {
+        checkDirectoryAlreadyExists(keyName, ozoneManager,
+                reachedLastPathComponent);
```

Comment:
```xml
  protected boolean checkDirectoryAlreadyExists(String volumeName,
      String bucketName, String keyName, OMMetadataManager omMetadataManager)
      throws IOException {
    if (omMetadataManager.getKeyTable().isExist(
        omMetadataManager.getOzoneDirKey(volumeName, bucketName,
            keyName))) {
      return true;
    }
    return false;
  }
```
Original logic is to check if there is the same name dir key already stored in key table.
For here,  the logic is that we should use new file key to check if that existed in file table, not check the dir table.
checkDirectoryAlreadyExists will always throw OMException error once OzoneFileSystem enabled under above logic, because parents dir is already created during File create request.

Reply:
Thanks for the comment. I think the logic is fine, but will discuss for better clarity and happy to fix if we finds any gap.

Yes, the existing logic in `OMKeyRequest#checkDirectoryAlreadyExists` checks the directory existence for the given committing fileName.

New logic also does the directory existence check against `DirTable` for the given committing fileName. `FileTable` stores only the file information <parentID/fileName> and will not know about the DirInfo.

For example,  committing fileName is `/dir1/dir2/dir3/file1`.

**Existing Logic:** 
step-1) It will add trailing slash - `/dir1/dir2/dir3/file1/` as the directory key entries represented by trailing slash.
step-2) Then check `/dir1/dir2/dir3/file1/` in the KeyTable. 

**New Logic:** 
step-1) Traverse till the leaf node. I have used a flag `boolean reachedLastPathComponent = !pathComponents.hasNext();` to mark that the traversal reached till end.
step-2) Once it reached the last path element, if it finds an OmDirectoryInfo from `DirTable` corresponding to the 'file1' then it throws OMException.
```

    omDirectoryInfo = omMetadataManager.
              getDirectoryTable().get(dbNodeName);
      if (omDirectoryInfo != null) {
        checkDirectoryAlreadyExists(keyName, ozoneManager,
                reachedLastPathComponent);
        lastKnownParentId = omDirectoryInfo.getObjectID();
      } 
```

Reply:
Thanks for the detailed explanation, I missed that fileName can also be passed during directory key entries check.